### PR TITLE
gh-144156: Fix email header folding concatenating encoded words

### PR DESCRIPTION
**Pull request by robsdedude**
* Issue: gh-144156

### CODE REVIEW

File: Lib/email/_header_value_parser.py [hunk 1]
```diff
@@ -2835,6 +2835,27 @@ def _steal_trailing_WSP_if_exists(lines):
             lines.pop()
     return wsp
 
+def _steal_all_trailing_WSP_if_exists(lines):
```
Comment (aisk) [inline Lib/email/_header_value_parser.py]: According to PEP8, there should be two blank lines here. L2858 should be the same.

File: Lib/test/test_email/test_generator.py [hunk 2]
```diff
@@ -393,6 +393,24 @@ def test_defaults_handle_spaces_at_start_of_continuation_line(self):
         g.flatten(msg)
         self.assertEqual(s.getvalue(), expected)
 
+    # gh-144156
+    # https://github.com/python/cpython/issues/144156
```
Comment (aisk) [inline Lib/test/test_email/test_generator.py]: The full URL link is redundant. I suggest using `# gh-144156: a brief description.`

File: Lib/email/_header_value_parser.py (hunk 1 above)
Comment (robsdedude) [reply Lib/email/_header_value_parser.py]: I shall fix. I was under the impression some linter or formatter were running in the pre-commit hooks or CI. While they are, non seem to be covering this file. I've basically copied the code-format of the surrounding code (most of the module-level functions in this file only have a single blank like separating them).

Comment (aisk) [reply Lib/email/_header_value_parser.py]: Some old code doesn’t follow PEP 8, but new code should.

**Comment by bitdancer** [comment]
FYI I've been looking at this and will have a review soonish.  You are correct that the  gh-92281 fix was buggy.

File: Lib/email/_header_value_parser.py [hunk 3]
```diff
@@ -2835,6 +2835,29 @@ def _steal_trailing_WSP_if_exists(lines):
...
+        wsp_lines.insert(0, wsp_line)
+        lines[-1] = lines[-1][:i]
+        if not lines[-1]:
+            lines_popped = True
+            lines.pop()
+        else:
+            break
+    if lines_popped:
+        lines.append(' ' if lines else '')
+    return ''.join(wsp_lines)
```
Comment (bitdancer) [inline Lib/email/_header_value_parser.py]: I don't think this is necessary, see my suggestion below.

File: Lib/email/_header_value_parser.py [hunk 4]
```diff
@@ -2843,9 +2866,7 @@ def _refold_parse_tree(parse_tree, *, policy):
     maxlen = policy.max_line_length or sys.maxsize
     encoding = 'utf-8' if policy.utf8 else 'us-ascii'
     lines = ['']  # Folded lines to be output
-    leading_whitespace = ''  # When we have whitespace between two encoded
-                             # words, we may need to encode the whitespace
-                             # at the beginning of the second word.
+    last_word_is_ew = False
     last_ew = None  # Points to the last encoded character if there's an ew on
```
Comment (bitdancer) [inline Lib/email/_header_value_parser.py]: This comment that was added by 90281 is incorrect.  It could say "if there is an encoded word in the last line of lines, points to the encoded word's first character".  Which means it is a badly named variable (that's my fault).  We could rename it 'last_ew_start', but I don't know if we want to do that in this PR.  The comment should be fixed, though.

File: Lib/email/_header_value_parser.py [hunk 5]
```diff
@@ -2917,10 +2940,8 @@ def _refold_parse_tree(parse_tree, *, policy):
                      last_charset == 'utf-8' and charset != 'us-ascii')):
                     last_ew = None
                 last_ew = _fold_as_ew(tstr, lines, maxlen, last_ew,
-                                      part.ew_combine_allowed, charset, leading_whitespace)
-                # This whitespace has been added to the lines in _fold_as_ew()
-                # so clear it now.
-                leading_whitespace = ''
+                                      part.ew_combine_allowed, charset, last_word_is_ew)
```
Comment (bitdancer) [inline Lib/email/_header_value_parser.py]: Let's just reformat this:
```
last_ew = _fold_as_ew(
    tstr,
    lines,
    maxlen,
    last_ew,
    part.ew_combine_allowed,
    charset,
    last_word_is_ew,
    )
```

File: Lib/email/_header_value_parser.py [hunk 6]
```diff
@@ -2933,28 +2954,20 @@ def _refold_parse_tree(parse_tree, *, policy):
 
         if len(tstr) <= maxlen - len(lines[-1]):
             lines[-1] += tstr
+            if any(char not in WSP for char in tstr):
+                last_word_is_ew = False
```
Comment (bitdancer) [inline Lib/email/_header_value_parser.py]:
```suggestion
                last_word_is_ew = not tstr.strip(_WSP) and last_word_is_ew
```

You'll need to add
```
_WSP = ''.join(WSP)
```

up at the top of the file where WSP is defined.  (I'm actually adding that constant myself in a different PR I'm working on).

I haven't run timeit on it, but this should be more efficient as well as more compact. Hmm.  I suppose it would be faster to have last_word_is_ew first, but then we should maybe wrap the tstr reference in bool for cleanliness.  Up to you.

File: Lib/email/_header_value_parser.py [hunk 7]
```diff
@@ -2933,28 +2954,20 @@ def _refold_parse_tree(parse_tree, *, policy):
...
                 lines.append(newline + tstr)
-
-                whitespace_accumulator = []
-                for char in lines[-1]:
-                    if char not in WSP:
-                        break
-                    whitespace_accumulator.append(char)
-                leading_whitespace = ''.join(whitespace_accumulator)
+                if any(char not in WSP for char in lines[-1]):
+                    last_word_is_ew = False
```
Comment (bitdancer) [inline Lib/email/_header_value_parser.py]:
```suggestion
                    last_word_is_ew = not tstr.strip(_WSP) and last_word_is_ew
```

File: Lib/email/_header_value_parser.py [hunk 8]
```diff
@@ -2994,10 +3007,12 @@ def _refold_parse_tree(parse_tree, *, policy):
         else:
             # We can't fold it onto the next line either...
             lines[-1] += tstr
+        if any(char not in WSP for char in tstr):
+            last_word_is_ew = False
```
Comment (bitdancer) [inline Lib/email/_header_value_parser.py]:
```suggestion
        last_word_is_ew = not tstr.strip(_WSP) and last_word_is_ew     
```

File: Lib/email/_header_value_parser.py [hunk 9]
```diff
@@ -3012,14 +3027,21 @@ def _fold_as_ew(to_encode, lines, maxlen, last_ew, ew_combine_allowed, charset,
...
         if (len(lines[-1]) == maxlen):
             lines.append(_steal_trailing_WSP_if_exists(lines))
         lines[-1] += leading_wsp
+    elif last_word_is_ew:
+        # If we are following up an encoded word with another encoded word,
+        # any white space between the two will be ignored when decoded.
+        # Therefore, we encode all to-be-displayed whitespace in the second
+        # encoded word.
+        leading_whitespace = _steal_all_trailing_WSP_if_exists(lines)
+        to_encode = leading_whitespace + to_encode
```
Comment (bitdancer) [inline Lib/email/_header_value_parser.py:3023]:
```suggestion
        len_without_wsp = len(lines[-1].rstrip(_WSP))
        leading_whitespace = lines[-1][len_without_wsp:]
        lines[-1] = lines[-1][:len_without_wsp] + (
            ' ' if leading_whitespace else '')
        to_encode = leading_whitespace + to_encode
```

This avoids that complicated single use function. We should never be producing lines that contain only blanks, they get encoded.

You can also move this up to be the first elif and drop the 'and not last_word_is_ew'.

File: Lib/email/_header_value_parser.py [hunk 10]
```diff
@@ -3040,20 +3062,11 @@ def _fold_as_ew(to_encode, lines, maxlen, last_ew, ew_combine_allowed, charset,
 
     while to_encode:
         remaining_space = maxlen - len(lines[-1])
-        text_space = remaining_space - chrome_len - len(leading_whitespace)
+        text_space = remaining_space - chrome_len
         if text_space <= 0:
             lines.append(' ')
             continue
```
Comment (bitdancer) [inline Lib/email/_header_value_parser.py:3057]: One of my new tests revealed a pre-existing problem here:
```suggestion
        if text_space <= 0:
            newline = _steal_trailing_WSP_if_exists(lines)
            lines.append(newline or ' ')
            new_last_ew = len(lines[-1])
            continue
```

File: Lib/test/test_email/test_generator.py [hunk 11]
```diff
@@ -393,6 +393,26 @@ def test_defaults_handle_spaces_at_start_of_continuation_line(self):
...
+            b" semaines =?utf-8?q?d=E2=80=99abonnement_=C3=A0?= 24 heures, Bilan, Tribune de\n"
+            b" =?utf-8?q?Gen=C3=A8ve?= et tous les titres Tamedia\n\n"
+        )
+        msg = EmailMessage()
+        msg['Subject'] = source
+        s = io.BytesIO()
+        g = BytesGenerator(s)
+        g.flatten(msg)
+        self.assertEqual(s.getvalue(), expected)
+
```
Comment (bitdancer) [inline Lib/test/test_email/test_generator.py:415]: Here is a test harness I used to verify the code:

```
    def test_ew_folding_round_trip_random(self):
        from random import randint
        c = 30
        b = 30
        w = 30
        source = ''
        #source += ' ' * randint(0, b)
        source += ('a' * randint(1, c), 'ф' * randint(1, c))[randint(0, 1)]
        for i in range(randint(1, w)):
            source += ' ' * randint(1, b)
            source += ('a' * randint(1, c), 'ф' * randint(1, c))[randint(0, 1)]
        source += ' ' * randint(0, b)
        with open('temp', 'a') as f:
            print(source, file=f)
        msg = EmailMessage()
        msg['Subject'] = source
        s = io.BytesIO()
        g = BytesGenerator(s, maxheaderlen=30)
        g.flatten(msg)
        flat = s.getvalue()
        reparsed = message_from_bytes(flat, policy=policy.default)['Subject']
        self.assertMultiLineEqual(reparsed, source)
```

I'm not sure if we should include a variation on this in the PR or not, but you might want to run it a bunch, maybe tweaking the constants, to see if I missed anything.  I ran it for a while with the above parameters and saw no failures.

Here are two tests that the above produced, that located the pre-existing bug I mentioned above:

```
    def test_ew_folding_round_trip_1(self):
        print()
        source = "aaaaaaaaa фффффффф "
        msg = EmailMessage()
        msg['Subject'] = source
        s = io.BytesIO()
        g = BytesGenerator(s, maxheaderlen=30)
        g.flatten(msg)
        flat = s.getvalue()
        reparsed = message_from_bytes(flat, policy=policy.default)['Subject']
        self.assertMultiLineEqual(reparsed, source)

    def test_ew_folding_round_trip_2(self):
        print()
        source = "aaa aaaaaaa   aaa ффф фффф  "
        msg = EmailMessage()
        msg['Subject'] = source
        s = io.BytesIO()
        g = BytesGenerator(s, maxheaderlen=30)
        g.flatten(msg)
        flat = s.getvalue()
        reparsed = message_from_bytes(flat, policy=policy.default)['Subject']
        self.assertMultiLineEqual(reparsed, source)
```

File: Misc/NEWS.d/next/Core_and_Builtins/2026-02-10-22-05-51.gh-issue-144156.UbrC7F.rst [hunk 12]
```diff
@@ -0,0 +1,3 @@
+Fix folding of email headers violating `RFC 2047`_ with two consecutive encoded words without separating linear-white-space.
+
+.. _RFC 2047: https://www.rfc-editor.org/rfc/rfc2047
```
Comment (bitdancer) [inline Misc/NEWS.d/next/Core_and_Builtins/2026-02-10-22-05-51.gh-issue-144156.UbrC7F.rst]:
```suggestion
Fix the folding of headers by the :mod:`email` library when :rfc:`2047` encoded words are used.  Now whitespace is correctly preserved and also correctly added between adjacent encoded words.  The latter property was broken by the fix for gh-92081, which mostly fixed previous failures to preserve whitespace.
```

Out of curiosity I checked, and according to the changelogs the buggy fix was released in 3.14 alpha 1, 3.13 beta 2 and 3.12.5.  Unfortunately the latter is now final.  Maybe we should include that in the news item?  I'm not sure what the policy is on that.

**Review by bitdancer — CHANGES_REQUESTED** [review]
Thanks for fixing this!

I edited this review a bunch, so mistake could have crept in to my code suggestions, though I did test them.  The randomizer test should catch any transcription errors if nothing else does.

File: Lib/email/_header_value_parser.py (hunk 6 above)
Comment (robsdedude) [reply Lib/email/_header_value_parser.py]: Much faster indeed.

```python
import timeit

_wSP = ' \t'
WSP = set(_wSP)

def one(s):
    return bool(s.strip(_wSP))

def two(s):
    return all(c in WSP for c in s)

def three(s):
    return not any(c not in WSP for c in s)

s1 = ' ' * 80
s2 = ' ' * 40 + 'a' + ' ' * 40
s3 = ' ' + 'a' * 80 + ' '

print('s1')
print(timeit.timeit(lambda: one(s1), number=10_000))
print(timeit.timeit(lambda: two(s1), number=10_000))
print(timeit.timeit(lambda: three(s1), number=10_000))

print('s2')
print(timeit.timeit(lambda: one(s2), number=10_000))
print(timeit.timeit(lambda: two(s2), number=10_000))
print(timeit.timeit(lambda: three(s2), number=10_000))

print('s3')
print(timeit.timeit(lambda: one(s3), number=10_000))
print(timeit.timeit(lambda: two(s3), number=10_000))
print(timeit.timeit(lambda: three(s3), number=10_000))
```

```
s1
0.014692971999920701
0.22478941200006375
0.22499246799998218
s2
0.01503182199985531
0.12194120699996347
0.1325552719999905
s3
0.011907245000202238
0.025579678999747557
0.025823422000030405
```

File: Lib/email/_header_value_parser.py (hunk 9 above)
Comment (robsdedude) [reply Lib/email/_header_value_parser.py:3023]: > We should never be producing lines that contain only blanks, they get encoded.

While testing, I noticed, too, that wrapped spaces are always encoded. TBH, I didn't like that and wanted to come up with a solution that's robust to changing this behavior in the future. I didn't like it because reading the RFC and the rest of this code it sure seems like not encoding words is generally preferable because of its human readability (and to a lesser extent mail clients that don't support encoded words). Further, I thought assuming an invariant that's enforced a few hundred lines away in a different function is a nice recipe for spooky action at a distance.

If you anyway think the fix should assume this invariant, I'll change this part.

File: Misc/NEWS.d/next/Core_and_Builtins/2026-02-10-22-05-51.gh-issue-144156.UbrC7F.rst (hunk 12 above)
Comment (robsdedude) [reply Misc/NEWS.d/next/Core_and_Builtins/2026-02-10-22-05-51.gh-issue-144156.UbrC7F.rst]: > Maybe we should include that in the news item?

I'm all for documenting known issues and shortcomings. However, I doubt people would find this info in the release notes unless they're doing "software archeology", i.e., they already know (roughly) what info they're looking for.

> I'm not sure what the policy is on that.

Neither am I. I'm pretty much a first-time contributor :sweat_smile:

File: Lib/email/_header_value_parser.py (hunk 6 above)
Comment (robsdedude) [reply Lib/email/_header_value_parser.py]: However, I find combining the if statement into a boolean expression (`last_word_is_ew = last_word_is_ew and not tstr.strip(_WSP)`) is much harder to read. At least for me. I needs some proper thinking about what the intention of that code was. I therefore decided to put that logic into a helper function.

File: Lib/email/_header_value_parser.py (hunk 10 above)
Comment (robsdedude) [reply Lib/email/_header_value_parser.py:3057]: This suggestion does not seem to fix the test cases `test_ew_folding_round_trip_1` and `test_ew_folding_round_trip_2` you posted in a comment below :thinking:

File: Lib/email/_header_value_parser.py (hunk 6 above)
Comment (bitdancer) [reply Lib/email/_header_value_parser.py]: Depends on how one's brain works, I guess.  I had to think about it to realize that the if was going to leave the flag True if it was already true, whereas my one liner was immediately clear to me ;)

File: Lib/email/_header_value_parser.py (hunk 9 above)
Comment (bitdancer) [reply Lib/email/_header_value_parser.py:3023]: Well, you have to keep in mind that the RFC requires that there be no line that consists of only whitespace.  In theory the folder is only supposed to encode whitespace that it has to, but a completely blank line is an example of "has to".  If it is being overly aggressive and encoding whitespace it doesn't need to that would be a bug.  Though as the need for this PR demonstrates, getting it right isn't easy ;)

As for the spooky action at a distance...yes, this code has grown into a spaghetti ball.  Believe it or, the original (pre bug fixing) version of this code was like the fourth iteration of ending up with an incomprehensible mess and starting over.  I have plans to redo the whole thing from scratch again, and do something like have an accumulator object that would own things like the last_word_is_ew flag, but I'm working on rewriting the parser first, and that is taking a while.  And I want to add a lot more tests before I attempt a rewrite, to make sure I don't break anything.

File: Lib/email/_header_value_parser.py (hunk 10 above)
Comment (bitdancer) [reply Lib/email/_header_value_parser.py:3057]: Hmm.  Maybe there's a bug in your steal all whitespace function?  Using my suggestion the tests pass for me, but I see you didn't make that change, at least in what I can see.  On the other hand, github also doesn't show the fix applied, so I'm not sure where you are at, exactly.

However, I also see that I failed to run the full test suite before posting the review, and there's a whitespace change in three existing tests.  I think it's a bugfix, but I'll have to look closer when I get a chance later today.

File: Misc/NEWS.d/next/Core_and_Builtins/2026-02-10-22-05-51.gh-issue-144156.UbrC7F.rst [hunk 13]
```diff
@@ -0,0 +1,3 @@
+Fix folding of email headers violating `RFC 2047`_ with two consecutive encoded words without separating linear-white-space.
+
```
Comment (bitdancer) [inline Misc/NEWS.d/next/Core_and_Builtins/2026-02-10-22-05-51.gh-issue-144156.UbrC7F.rst]: Looks like this suggestion didn't apply cleanly.

File: Lib/email/_header_value_parser.py (hunk 9 above)
Comment (robsdedude) [reply Lib/email/_header_value_parser.py:3023]: > you have to keep in mind that the RFC requires that there be no line that consists of only whitespace.

Thanks, I overlooked that part in the RFC. In this case I agree that there's much less risk I will change the PR accordingly.

File: Lib/email/_header_value_parser.py (hunk 10 above)
Comment (robsdedude) [reply Lib/email/_header_value_parser.py:3057]: Yeah, I only had the change applied locally. I think I found the issue. The suggestion here looks like the `lines.append(' ')` right after `if text_space <= 0:` should remain there. I suppose it shouldn't. Makes more sense that way when reading the code, too :upside_down_face:

With the suggestion applied correctly to my local working tree I now see the same: your new tests pass, 3 old tests started failing.

I went ahead and adjusted the test expectations. Let me know what you think once you've had a chance to have a closer look.

File: Lib/email/_header_value_parser.py (hunk 3 above)
Comment (robsdedude) [reply Lib/email/_header_value_parser.py]: Removed, as discussed.

File: Lib/email/_header_value_parser.py [hunk 14]
```diff
@@ -2835,6 +2836,13 @@ def _steal_trailing_WSP_if_exists(lines):
             lines.pop()
     return wsp
 
+
+def _last_word_is_sill_ew(_last_word_is_ew, added_str):
```
Comment (bitdancer) [inline Lib/email/_header_value_parser.py]: I find this name a bit confusing.  What it actually does is "return true if last_word_is_ew is true and added_str consists of whitespace only".  I can't think of a name that is shorter and as descriptive as the actual code.  So I'm not in favor of this function.

File: Lib/email/_header_value_parser.py (hunk 9 above)
Comment (bitdancer) [reply Lib/email/_header_value_parser.py:3023]: I think it is better to move this elif up above the one it is currently after.  It only saves 23 characters, but since we can, why not :)  More importantly, the logic is IMO slightly easier to follow: first we check if we are appending to an encoded word, and then if not we handle the remaining special case that only matters when we aren't appending to an encoded word.

File: Lib/email/_header_value_parser.py (hunk 10 above)
Comment (bitdancer) [reply Lib/email/_header_value_parser.py:3057]: Yeah that suggestion looks borked.  I don't see any way to dismiss it, unfortunately.

The test expectation changes are correct, that extra space after Subject was a bug.  I may even have opened an issue about it, but if so I can't find it.  Probably just made a "note to self" that I've since lost.

**Review by bitdancer — CHANGES_REQUESTED** [review]

File: Lib/email/_header_value_parser.py (hunk 14 above)
Comment (robsdedude) [reply Lib/email/_header_value_parser.py]: Alright. Inlined it again.

File: Lib/email/_header_value_parser.py (hunk 9 above)
Comment (robsdedude) [reply Lib/email/_header_value_parser.py:3023]: Oh sorry. That last comment slipped my attention. Fixed now.

**Review by bitdancer — APPROVED** [review]

File: Misc/NEWS.d/next/Core_and_Builtins/2026-02-10-22-05-51.gh-issue-144156.UbrC7F.rst
Comment (StanFromIreland) [inline Misc/NEWS.d/next/Core_and_Builtins/2026-02-10-22-05-51.gh-issue-144156.UbrC7F.rst:1]: @robsdedude can you please open a PR to move this to the Library section.

Comment (robsdedude) [reply Misc/NEWS.d/next/Core_and_Builtins/2026-02-10-22-05-51.gh-issue-144156.UbrC7F.rst:1]: *  dev: gh-145205
  * 3.14: gh-145207
  * 3.13: for this one it seems bitdancer has already fixed it in the the initial PR

Comment (StanFromIreland) [reply Misc/NEWS.d/next/Core_and_Builtins/2026-02-10-22-05-51.gh-issue-144156.UbrC7F.rst:1]: Oh you didn't have to open two, we can auto back port from main, thanks!

Comment (robsdedude) [reply Misc/NEWS.d/next/Core_and_Builtins/2026-02-10-22-05-51.gh-issue-144156.UbrC7F.rst:1]: 😅 I'll know for next time. Thank you 🙇
